components/supervisor/pkg/serverapi/metrics.go

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License.AGPL.txt in the project root for license information.
+
+package serverapi
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	ServerTypeServerAPI = "server-api"
+	ServerTypePublicAPI = "public-api"
+)
+
+type ClientMetrics struct {
+	clientHandledCounter   *prometheus.CounterVec
+	clientHandledHistogram *prometheus.HistogramVec
+}
+
+func NewClientMetrics() *ClientMetrics {
+	return &ClientMetrics{
+		clientHandledCounter: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "supervisor_client_handled_total",
+				Help: "Total number of supervisor outgoing services completed by the client, regardless of success or failure.",
+				// add server label to identify it's going for server api or public api
+			}, []string{"method", "server", "err_code"}),
+		clientHandledHistogram: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name: "supervisor_client_handling_seconds",
+				Help: "Histogram of response latency (seconds) of the supervisor outgoing services until it is finished by the application.",
+				// it should be aligned with https://github.com/gitpod-io/gitpod/blob/84ed1a0672d91446ba33cb7b504cfada769271a8/install/installer/pkg/components/ide-metrics/configmap.go#L315
+				Buckets: []float64{0.1, 0.2, 0.5, 1, 2, 5, 10},
+			}, []string{"method", "server", "err_code"}),
+	}
+}
+
+func (c *ClientMetrics) ProcessMetrics(usePublicAPI bool, method string, err error, startTime time.Time) {
+	code := status.Code(err)
```

Comment:
what happens if err is not grpc error, i.e from server api or some other system error? Should not we map it to something like `Unknown` or proper JSON RPC error code, e.g. 401?

Reply:
`status.Code` maps it to unkown intenally

Reply:
yeah, i was wondering whether we would like to gain insights from server side as well, maybe we can see that public api already better and then can push forward with it?

Could we reuse https://github.com/gitpod-io/gitpod/blob/df0872d030503c6d79a9fd8b7d4775adebcf2632/components/public-api-server/pkg/proxy/errors.go#L17 here?